# Task 4.6 

Content:

1.- Importing Libraries and Dataset

2.- Checking the characteristics of the imported dataframes

3.- Merging both Dataframe



## 1.- Importing libraries and data

In [9]:
import pandas as pd
import numpy as np
import os

path = r'C:\Users\cmfag\OneDrive\Desktop\Data Analyst Specialist\Achievement 4\4.3_orders_products\02 Data\Prepared Data'

df_orders_products_combined = pd.read_pickle(os.path.join(path, 'orders_products_combined.pkl'))

# Importing product wrangled, cleaned, and deduped 


df_prods = pd.read_csv(os.path.join(path, 'products_checked.csv'), index_col = False)




## 2.- Checking the characteristics of the imported dataframes 

### 2.1- Checking the shape of: df_orders_products_combined

In [12]:
df_orders_products_combined.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [13]:
df_orders_products_combined.shape

(32434489, 11)

In [14]:
df_orders_products_combined.describe()

order_id       user_id  order_number     order_dow  \
count  3.243449e+07  3.243449e+07  3.243449e+07  3.243449e+07   
mean   1.710749e+06  1.029372e+05  1.714205e+01  2.738818e+00   
std    9.873007e+05  5.946648e+04  1.753504e+01  2.090049e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.559430e+05  5.142100e+04  5.000000e+00  1.000000e+00   
50%    1.711048e+06  1.026110e+05  1.100000e+01  3.000000e+00   
75%    2.565514e+06  1.543910e+05  2.400000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.243449e+07            3.035642e+07  3.243449e+07   
mean        1.342498e+01            1.110407e+01  2.557634e+04   
std         4.246365e+00            8.778914e+00  1.409669e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.353000e+04   
50%         1.300000e+01            8.000000e+00  2.525600e+04   
75%         1.600000e+01            1.500000e+01  3.793500e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered  
count       3.243449e+07  3.243449e+07  
mean        8.351076e+00  5.896975e-01  
std         7.126671e+00  4.918886e-01  
min         1.000000e+00  0.000000e+00  
25%         3.000000e+00  0.000000e+00  
50%         6.000000e+00  1.000000e+00  
75%         1.100000e+01  1.000000e+00  
max         1.450000e+02  1.000000e+00

### 2.2 Checking the shape of: df_prods

In [16]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [17]:
df_prods.shape

(49672, 6)

### Removing _merge columns to avoid any error in later steps.

In [19]:
df_orders_products_combined = df_orders_products_combined.drop(columns=['_merge'], errors='ignore')
df_prods = df_prods.drop(columns=['_merge'], errors='ignore')

df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [20]:
df_prods.describe()

Unnamed: 0    product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000  49672.000000   49672.000000  49672.000000
mean   24852.005053  24850.349775     67.762442      11.728942      9.993282
std    14342.265579  14340.705287     38.315784       5.850779    453.615536
min        0.000000      1.000000      1.000000       1.000000      1.000000
25%    12432.750000  12432.750000     35.000000       7.000000      4.100000
50%    24851.500000  24850.500000     69.000000      13.000000      7.100000
75%    37272.250000  37268.250000    100.000000      17.000000     11.100000
max    49692.000000  49688.000000    134.000000      21.000000  99999.000000

In [21]:
df_orders_products_combined.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  
0                     NaN         196                  1          0  
1                     NaN       14084                  2          0  
2                     NaN       12427                  3          0  
3                     NaN       26088                  4          0  
4                     NaN       26405                  5          0

## 3.- Merging both Dataframe

### Suitable way to combine both dataframe 


Possible merge of both dataframes by using product_id as key variable

suitable merge form: left

After merge: verify result.



In [24]:
df_ords_prods_merge = df_orders_products_combined.merge(df_prods, on='product_id')

df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0         195                                     Soda        77   
1       14084  Organic Unsweetened Vanilla Almond Milk        91   
2       12427                      Original Beef Jerky        23   
3       26089               Aged White Cheddar Popcorn        23   
4       26406         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices  
0              7     9.0  
1             16    12.5  
2             19     4.4  
3             19     4.7  
4             17     1.0

Confirming the Results of the Merge Using a Merge Flag

In [26]:

df_ords_prods_merge = df_orders_products_combined.merge(df_prods, on='product_id', indicator=True)

df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0         195                                     Soda        77   
1       14084  Organic Unsweetened Vanilla Almond Milk        91   
2       12427                      Original Beef Jerky        23   
3       26089               Aged White Cheddar Popcorn        23   
4       26406         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices _merge  
0              7     9.0   both  
1             16    12.5   both  
2             19     4.4   both  
3             19     4.7   both  
4             17     1.0   both

In [27]:
# Changing the name of the merge column to avoid conflicts in the future!
df_ords_prods_merge = df_orders_products_combined.merge(df_prods, on='product_id', how='left', indicator='merge_check')

In [28]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check  
0            7.0     9.0        both  
1           16.0    12.5        both  
2           19.0     4.4        both  
3           19.0     4.7        both  
4           17.0     1.0        both

In [29]:
df_ords_prods_merge['merge_check'].value_counts()

merge_check
both          32404859
left_only        30200
right_only           0
Name: count, dtype: int64

In [42]:
df_ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

## Conclusions out of the Merging Process

Based on the merge_check counts 
32,404,859 rows where products_id mached in both df_orders_products_combined and df_prods. 
Only 30,200 rows did not match in df_prods. This could be expected if there are some products in orders_products_combined that are missing or don’t exist in df_prods.

## Export the dataframe df_ords_prods_merge

In [ ]:
df_ords_prods_merge.to_pickle(os.path.join(path, 'ords_prods_merge.pkl'))